### ISSUE TITLE
Printer : unable to print large files

### ISSUE THREAD

**Issue by rgc2000** [GideonZ/1541ultimate#342]
A change in the IEC poll method causes the IEC processor to be disabled while printing large files (with lots of IEC trafic) and I have the ? DEVICE NOT PRESENT ERROR message in the middle of a print job.

In the debug console I can see the message "-> ** IEC Processor seems to have gone haywire. Stopped! **" Depending on the cartridge this happens sooner (on U2+) or later (on U2). I need to comment out the looplimit code that was added recently to be able to print again until the job is done.

**Comment by GideonZ** [comment]
Hehe... Oh god.

**Comment by GideonZ** [comment]
It is an active polling, but with a task delay in the loop. Many mechanisms
are derived from the event driven implementation of firmware 2.x. There are
still more instances in the firmware where a task is created that calls the
old "poll" function repeatedly.

I am surprised that the FIFO status is actually non-empty for such a long
time. You see, I would expect that the CPU is fast enough to take at most
some chars from the FIFO and then retire, since IEC is slow.

I think the problem is solved completely when the data from the FIFO is
pushed into a queue to the printer handler. Maybe there should be a special
control character to keep the IEC bus 'busy' if a queue is full, but the
data should not remain in the FIFO. That is not what it was meant for.

**Comment by GideonZ** [comment]
I propose to implement the loop limit as follows:

```
        uint8_t a;
        for (int loopcnt = 0; loopcnt < 500; loopcnt++) {
            a = HW_IEC_RX_FIFO_STATUS;
            if (a & IEC_FIFO_EMPTY) {
                break;
            }
```

This would limit the number of bytes transferred in one "poll" to 500, but it will always return to the TaskDelay after that to allow other threads get time. What do you think?

**Comment by rgc2000** [comment]
I'm going to test this on my two cartridges. I can see now how it works and the place where the scheduler can switch to other tasks. I was very surprised about the active polling in a multitask OS. I also understand why the cartridge is unresponsive while printing. The IEC bus can be very busy while printing, the C64 floods the printer and the whole job can be done in the IEC FIFO loop without breaking to previous infinite loop.
The PNG generation is still a very time consuming procedure and there is no task switching during this operation. That's why I added a modal message with a progress bar if it is started by the action menu "printer - flush/eject".

**Comment by rgc2000** [comment]
I confirm that this solution works perfectly. The printer is now able to print large files while the rest of the software can have access from time to time to the CPU. I can see it on the activity LED witch is now blinking periodically while printing (while LED off the printer has lost CPU). This has been tested with 3.10h on both U2 and U2+. 10 pages printed in a row with no issue.

I have still an issue on U2+ about activity LED which is off while generating PNG data. I have to figure out why but this is another topic.

**Comment by rgc2000** [comment]
Fixed in firmware 3.11
